File: flutter/lib/core/models/kb.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'kb.freezed.dart';
part 'kb.g.dart';

enum KbKind {
  fact,
  paragraph;

  String get label => switch (this) {
    KbKind.fact => '事实',
    KbKind.paragraph => '段落',
  };
}

/// 知识库检索命中的一条事实或段落。
@freezed
abstract class KbHit with _$KbHit {
  const factory KbHit({
    required KbKind kind,
    @Default('') String pmid,
    @Default('') String doi,
    @Default('') String pmcid,
    @Default('') String title,
    @Default('') String year,
    @Default('') String journal,
    @Default('') String quartile,
    @Default('') String source,
    @Default('') String authors,
    int? pid,
    String? sec,
    int? page,
    @Default('') String text,
    String? textZh,
    String? factKind,
    String? quote,
    bool? verified,
    @Default(0) double score,
  }) = _KbHit;

  factory KbHit.fromJson(Map<String, Object?> json) => _$KbHitFromJson(json);
}

@freezed
abstract class KbSearchResult with _$KbSearchResult {
  const factory KbSearchResult({
    @Default('') String query,
    @Default(<KbHit>[]) List<KbHit> items,
  }) = _KbSearchResult;

  factory KbSearchResult.fromJson(Map<String, Object?> json) =>
      _$KbSearchResultFromJson(json);
}

@freezed
abstract class KbStats with _$KbStats {
  const factory KbStats({
    @Default(0) int items,
    @Default(0) int papers,
    @Default(<String, int>{}) Map<String, int> byKind,
    String? embedder,
    int? dim,
  }) = _KbStats;

  factory KbStats.fromJson(Map<String, Object?> json) =>
      _$KbStatsFromJson(json);
}

File: apple/PicoSeekKit/Sources/PicoSeekKit/Models/Kb.swift
import Foundation

public enum KbKind: String, Codable, Sendable, Hashable, CaseIterable {
    case fact
    case paragraph

    public var label: String {
        switch self {
        case .fact: "事实"
        case .paragraph: "段落"
        }
    }
}

/// 知识库检索命中的一条事实或段落。
public struct KbHit: Codable, Sendable, Hashable {
    public var kind: KbKind
    public var pmid: String
    public var doi: String
    public var pmcid: String
    public var title: String
    public var year: String
    public var journal: String
    public var quartile: String
    public var source: String
    public var authors: String
    public var pid: Int?
    public var sec: String?
    public var page: Int?
    public var text: String
    public var textZh: String?
    public var factKind: String?
    public var quote: String?
    public var verified: Bool?
    public var score: Double
}

public struct KbSearchResult: Codable, Sendable, Hashable {
    public var query: String
    public var items: [KbHit]
}

public struct KbStats: Codable, Sendable, Hashable {
    public var items: Int
    public var papers: Int
    public var byKind: [String: Int]
    public var embedder: String?
    public var dim: Int?
}
